### ISSUE TITLE
Feature request: Save optimizer variables under separate name scope

### ISSUE THREAD

**Issue by tsoernes** [tensorflow/tensorflow#16781]
Tensorflow version b'v1.3.0-rc1-3011-gd86448938' 1.3.0
Have I written custom code N/A
OS Platform and Distribution N/A 
TensorFlow installed from N/A
Bazel version N/A
CUDA/cuDNN version N/A
GPU model and memory N/A
Exact command to reproduce N/A

The variables initialized by the Adam optimizer is not created under its own name scope, but inherits the name of trainable variables, e.g. `model/q_networks/online/conv2d/bias/Adam_1` (the Adam optimizer was not creater under the name scopes `model`, `q_networks` or `online`).

This creates an issue if you want to use another optimizer for a restored model, because you can't use the name scope to avoid restoring the variables related to the optimizer.

Here's a workaround:
```
        build_model()  # Build model under namescope 'model', without optimizer
        temp = set(tf.global_variables())
        if save or restore:
            saver = tf.train.Saver(
                var_list=tf.get_collection(tf.GraphKeys.GLOBAL_VARIABLES, scope='model'))
        sess.run(tf.global_variables_initializer())
        if restore:
            saver.restore(sess, model_path)
        trainer = tf.train.AdamOptimizer(1e-4)
        sess.run(tf.variables_initializer(set(tf.global_variables()) - temp))  # Initialize Adam variables
```
It would be simpler (to figure out) if the optimizer variables had their own scope.

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tsoernes** [comment]
Yes.

**Comment by bignamehyp** [comment]
We will consider it if we heard more demand for this feature.

**Comment by themightyoarfish** [comment]
I face this issue when trying to use weights from some model as part for a larger one. I need to manually filter out the optimizer variables from the scope in which the variables which should be restored live.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by themightyoarfish** [comment]
Yes, this is still an issue.

**Comment by tsoernes** [comment]
yes

2018-04-17 14:42 GMT+02:00 Alfred Sorten Wolf <notifications@github.com>:

**Comment by martinwicke** [comment]
I believe we would have trouble changing this behavior before 2.0 since this would be a breaking change. So sadly, I think we're stuck with your workaround until then.

@allenlavoie has dealt with this particular issue before and correct me if I'm wrong.

**Comment by allenlavoie** [comment]
As of TensorFlow 1.5 you can use `optimizer.variables()` to get a list of variables associated with the optimizer. For Adam this is its `m` and `v` slot variables along with `beta1_power` and `beta2_power`. Note that this only works after a training step has been built, since optimizers create these variables lazily.

I think this works for the use-case you've described? You can use `.variables()` to filter the `var_list` passed to a `Saver` for restore, and you can use it to initialize the variables.

```
import tensorflow as tf

v = tf.get_variable(name="v", shape=[], use_resource=True)
optimizer = tf.train.AdamOptimizer(0.001)
train_op = optimizer.minimize(lambda: v.read_value())
print("All variables", tf.global_variables())
print("Non-optimizer variables",
      set(tf.global_variables()) - set(optimizer.variables()))
```
Prints:
```
All variables [<tf.Variable 'v:0' shape=() dtype=float32>, <tf.Variable 'beta1_power:0' shape=() dtype=float32_ref>, <tf.Variable 'beta2_power:0' shape=() dtype=float32_ref>, <tf.Variable 'v/Adam:0' shape=() dtype=float32>, <tf.Variable 'v/Adam_1:0' shape=() dtype=float32>]
Non-optimizer variables {<tf.Variable 'v:0' shape=() dtype=float32>}
```
This should work for any optimizer.

**Comment by tsoernes** [comment]
Good enough for me. Thanks!

2018-04-25 18:33 GMT+02:00 Allen Lavoie <notifications@github.com>:

**Comment by allenlavoie** [comment]
Cool. Will close for now, but feel free to comment if there are related issues.

Longer-term we'd like names not to matter as much, so getting variables from an object is slightly closer to that. [Object-based checkpointing](https://github.com/tensorflow/tensorflow/blob/8fa93c0eb6e12438c3fa564698c0bbf2921def9b/tensorflow/python/training/checkpointable_utils.py#L827) (not yet in core for a release, but it's been around as `tf.contrib.eager.Checkpoint` for a bit) should make things like this easier too, since you can just skip a dependency on the Optimizer object.
